README.rst
pythonz: a Python installation manager
======================================

Overview
--------

pythonz is a program to automate the building and installation of Python in the users $HOME. This is
a fork of the original project, `pythonbrew <https://github.com/utahta/pythonbrew>`_.

The original project seems to be unmaintained, and it also has some extra features which I don't really
need, so I made this for to make something a bit simpler that works for *me*. You may also find it
useful.

CPython, Stackless, PyPy and Jython are supported.

Installation
------------

The recommended way to download and install pythonz is to run these statements in your shell::

  curl -kL https://raw.github.com/saghul/pythonz/master/pythonz-install | bash

or::

  fetch -o - https://raw.github.com/saghul/pythonz/master/pythonz-install | bash

After that, pythonz installs itself to ``~/.pythonz``.

Please add the following line to the end of your ``~/.bashrc``::

  [[ -s $HOME/.pythonz/etc/bashrc ]] && source $HOME/.pythonz/etc/bashrc

If you need to install pythonz into somewhere else, you can do that by setting a ``PYTHONZ_ROOT`` environment variable::

  export PYTHONZ_ROOT=/path/to/pythonz
  curl -kLO https://raw.github.com/saghul/pythonz/master/pythonz-install
  chmod +x pythonz-install
  ./pythonz-install

For Systemwide (Multi-User) installation
----------------------------------------

If the install script is run as root, pythonz will automatically install into ``/usr/local/pythonz``.

pythonz will be automatically configured for every user on the system if you install it as root.

After installing it, where you would normally use `sudo`, non-root users will need to use `sudo-pythonz`::

  sudo-pythonz install 2.7.3

Before installing Pythons via Pythonz
-------------------------------------

You might want to install some optional dependencies, for functionality that
is often expected to be included in a Python build (it can be a bummer to discover these missing and
have to rebuild your python setup). These include the following, ordered by (very roughly guessed)
probability that you will need them::

Debian family (Ubuntu...)
^^^^^^^^^^^^^^^^^^^^^^^^^

::

  sudo apt-get install build-essential zlib1g-dev libbz2-dev libssl-dev libreadline-dev libncurses5-dev libsqlite3-dev libgdbm-dev libdb-dev libexpat-dev libpcap-dev liblzma-dev libpcre3-dev

If you need tkinter support, add **tk-dev**.

RPM family (CentOS, RHEL...)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^

::

  yum groupinstall "Development tools"
  yum install zlib-devel bzip2-devel openssl-devel readline-devel ncurses-devel sqlite-devel gdbm-devel db4-devel expat-devel libpcap-devel xz-devel pcre-devel

If you need tkinter support, add **tk-devel**.

OSX
^^^^^^^^^^^^^^^^^^^^^^^^^^^^

::

  xcode-select --install

Usage
-----

::

  pythonz command [options] version

See the available commands
^^^^^^^^^^^^^^^^^^^^^^^^^^

::

  pythonz help

To get help on each individual command
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

::

  pythonz help <command>


Install some pythons
^^^^^^^^^^^^^^^^^^^^

::

  pythonz install 2.7.3
  pythonz install -t stackless 2.7.2
  pythonz install -t jython 2.5.2
  pythonz install -t pypy --url https://bitbucket.org/pypy/pypy/downloads/pypy-1.8-osx64.tar.bz2 1.8
  pythonz install --verbose 2.7.2
  pythonz install --configure="CC=gcc_4.1" 2.7.2
  pythonz install --url http://www.python.org/ftp/python/2.7/Python-2.7.2.tgz 2.7.2
  pythonz install --file /path/to/Python-2.7.2.tgz 2.7.2
  pythonz install 2.7.3 3.2.3
  pythonz install -t pypy3 2.3.1

List the installed pythons
^^^^^^^^^^^^^^^^^^^^^^^^^^

::

  pythonz list

List all the available python versions for installing
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

::

  pythonz list -a

List all the available jython versions
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

::

   pythonz list -a -t jython

Uninstall the specified python
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

::

  pythonz uninstall 2.7.3
  pythonz uninstall -t stackless 3.2.2

Remove stale source folders and archives
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

::

  pythonz cleanup

Upgrade pythonz to the latest version
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

::

  pythonz update

Check the installed pythonz version
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

::

  pythonz version

Print the path to the interpreter of a given version
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

::

  pythonz locate 2.7.7

Recommended way to use a pythonz-installed version of Python
------------------------------------------------------------

For Python <= 3.2
^^^^^^^^^^^^^^^^^

Use `virtualenv`, e.g.::

  mkvirtualenv -p $(pythonz locate 2.7.3) python2.7.3

For more information about virtualenv, checkout `its documentation <http://www.virtualenv.org/en/latest/>`_.

For Python >= 3.3
^^^^^^^^^^^^^^^^^

Use `pyvenv` directly from Python, e.g.::

  /usr/local/pythonz/pythons/CPython-3.4.1/bin/pyvenv pyvenv

For more information about pyvenv, checkout `its documentation <https://docs.python.org/3/library/venv.html>`_.

DTrace support
--------------

CPython versions 2.7.6 and 3.3.4 can be built with DTrace suport. Patches adding support
for DTrace have been taken from `this page <http://www.jcea.es/artic/solitaire.htm/python_dtrace.htm>`_
by Jesús Cea.

Building Python with DTrace support::

  pythonz install --configure="--with-dtrace" 2.7.6


ChangeLog

2.0.1
=====

* Add support for CPython 3.5.2 and 2.7.12
* Add missing hash for CPython 3.4.4
* Add support for PyPy 5.3.0 and 5.3.1
* Add missing changelog entries

2.0.0
=====
* Disable progress bar if size unknown
* Merge pull request #105 from inirudebwoy/fix_81
* size could be None
* Fix list command for Python3 and tweak its logic
* Merge pull request #106 from berdario/fix_81
* List command: only use exact match
* Merge pull request #107 from berdario/fix_list
* bypass snvrevision and Makefile problem
* Merge pull request #112 from hackaugusto/issue109
* Remove setup.py and MANIFEST.in
* Remove warning when SHA256 hash is not available
* Fix checksum for Stackless 3.3.5
* Fix using --file on Python 3
* Add support for PyPy 5.X
* Add support for CPython 3.4.4

1.11.0
======
* Remove spurious install_dir checks and use a cleaner Exception when it already exists
* Add sha256sum for every python package, and supporting code to check it before installation
* Add ability to make external installations
* Add support for PyPy 4.0.0 and 4.0.1
* Use resumable-urlretrieve to resume interrupted downloads
* Add list -a filtering by groups
* Add support for CPython 2.7.11 and 3.5.1

1.10.0
======
* Add support for CPython 3.5.0
* Add support for PyPy 2.6.1

1.9.1
=====
* Add missing 'import sys' in locate command

1.9.0
=====
* Add support for PyPy 2.5.1 and 2.6.0
* Fix parsing versions of the form X.Y.ZZ
* Changed locate command to error out when a version is not found
* Add support for CPython 2.7.10 (6 days ago) <Chip Warden>
* Add support for Jython 2.7

1.8.0
=====
* Exit pythonz-install with non-zero status
* Fix system-wide installation
* Add support for CPython 3.4.3
* Add support for CPython 3.3.6
* Add install instructions for OS X

1.7.1
=====
* Add support for PyPy 2.5.0

1.7.0
=====
* Add support for CPython 2.7.9
* Make static builds by default on OSX too

1.6.1
=====
* Add support for PyPy3 2.4.0
* Simplify detecting file type
* Fix locate command

1.6.0
=====
* Add support for CPython 3.4.2
* Add support for PyPy 2.4.0
* Improved installation instructions
* Add -C and --configure options to bash completion
* Remove development branch

1.5.1
=====
* Add support for CPython 2.7.8

1.5.0
=====
* Add support for Stackless 3.2.5 and 3.3.5
* Add ability to reinstall an already installed version

1.4.1
=====
* Add support for PyPy3 2.3.1

1.4.0
=====
* Added locate command
* Add support for CPython 2.7.7
* Add support for PyPy 2.3.1
* Fixed creating bin/python symlink
* Update bash completion script for new commands

1.3.0
=====
* Add support for CPython 3.4.1
* Added PyPy 2.3 support
* Added support for Stackless 3.2.5 and 3.3.5
* Don't create unnecessary symlinks
* Properly patch Python 2.6.0 through Python 2.6.6

1.2.2
=====
* Added support for CPython 3.4.0

1.2.1
=====
* Initialize Python installation and build paths
* Removed executable permission from patch files

1.2.0
=====
* Added support for CPython 3.3.5
* Add mention of optional dependencies before build
* Use setuptools and fallback to distutils
* Add 'pythonz' entrypoint, useful for system-wide installations
* Simplified setting PATH environment variable

1.1.0
=====
* Added support for CPython 3.3.4
* Added DTrace patches for CPython 2.7.6 and 3.3.4 on OSX

1.0.0
=====
* First tagged release
* Support for CPython, PyPy, Stackless and Jython



pythonz/util.py

import os
import sys
import errno
import shutil
import re
import posixpath
import tarfile
import platform
import subprocess
import shlex
import select

PY3K = sys.version_info >= (3,)
if not PY3K:
    from urllib import quote as urlquote, unquote as urlunquote
    from urllib2 import urlparse
else:
    from urllib import parse as urlparse
    from urllib.parse import quote as urlquote, unquote as urlunquote

from pythonz.define import PATH_PYTHONS
from pythonz.log import logger


def is_url(name):
    try:
        result = urlparse.urlparse(name)
    except Exception:
        return False
    else:
        return result.scheme in ('http', 'https', 'file', 'ftp')

def is_file(name):
    try:
        result = urlparse.urlparse(name)
    except Exception:
        return False
    else:
        return result.scheme == 'file'

def splitext(name):
    base, ext = os.path.splitext(name)
    if base.lower().endswith('.tar'):
        ext = base[-4:] + ext
        base = base[:-4]
    return base, ext

def is_archive_file(name):
    ext = splitext(name)[1].lower()
    return ext in ('.zip', '.tar.gz', '.tar.bz2', '.tgz', '.tar')

def is_html(content_type):
    return content_type and content_type.startswith('text/html')

def is_gzip(content_type, filename):
    return content_type == 'application/x-gzip' or tarfile.is_tarfile(filename) or splitext(filename)[1].lower() in ('.tar', '.tar.gz', '.tgz', '.tar.bz2', '.tbz')

def get_macosx_deployment_target():
    m = re.search('^([0-9]+\.[0-9]+)', platform.mac_ver()[0])
    if m:
        return m.group(1)
    return None

def _py_version_cmp(v, v1, v2):
    if is_str(v):
        v = Version(v)
    return v >= v1 and v < v2

def is_python24(version):
    return _py_version_cmp(version, '2.4', '2.5')

def is_python25(version):
    return _py_version_cmp(version, '2.5', '2.6')

def is_python26(version):
    return _py_version_cmp(version, '2.6', '2.7')

def is_python27(version):
    return _py_version_cmp(version, '2.7', '2.8')

def is_python30(version):
    return _py_version_cmp(version, '3.0', '3.1')

def is_python31(version):
    return _py_version_cmp(version, '3.1', '3.2')

def is_python32(version):
    return _py_version_cmp(version, '3.2', '3.3')

def is_python33(version):
    return _py_version_cmp(version, '3.3', '3.4')

def makedirs(path):
    if not os.path.exists(path):
        os.makedirs(path)

def symlink(src, dst):
    try:
        os.symlink(src, dst)
    except OSError:
        pass

def unlink(path):
    try:
        os.unlink(path)
    except OSError as e:
        if e.errno != errno.ENOENT:
            raise

def rm_r(path):
    """like rm -r command."""
    if os.path.isdir(path):
        shutil.rmtree(path)
    elif os.path.isfile(path):
        unlink(path)

def split_leading_dir(path):
    path = str(path)
    path = path.lstrip('/').lstrip('\\')
    if '/' in path and (('\\' in path and path.find('/') < path.find('\\')) or '\\' not in path):
        return path.split('/', 1)
    elif '\\' in path:
        return path.split('\\', 1)
    else:
        return path, ''

def has_leading_dir(paths):
    """Returns true if all the paths have the same leading path name
    (i.e., everything is in one subdirectory in an archive)"""
    common_prefix = None
    for path in paths:
        prefix, rest = split_leading_dir(path)
        if not prefix:
            return False
        elif common_prefix is None:
            common_prefix = prefix
        elif prefix != common_prefix:
            return False
    return True

def untar_file(filename, location):
    if not os.path.exists(location):
        os.makedirs(location)
    if filename.lower().endswith('.gz') or filename.lower().endswith('.tgz'):
        mode = 'r:gz'
    elif filename.lower().endswith('.bz2') or filename.lower().endswith('.tbz'):
        mode = 'r:bz2'
    elif filename.lower().endswith('.tar'):
        mode = 'r'
    else:
        logger.error('Cannot determine compression type for file %s' % filename)
        mode = 'r:*'
    tar = tarfile.open(filename, mode)
    try:
        # note: python<=2.5 doesnt seem to know about pax headers, filter them
        leading = has_leading_dir([
            member.name for member in tar.getmembers()
            if not member.name.startswith('.') and member.name != 'pax_global_header'
        ])
        for member in tar.getmembers():
            fn = member.name
            if fn == 'pax_global_header':
                continue
            if leading:
                fn = split_leading_dir(fn)[1]
            path = os.path.join(location, fn)
            if member.isdir():
                if not os.path.exists(path):
                    os.makedirs(path)
            else:
                try:
                    fp = tar.extractfile(member)
                except (KeyError, AttributeError):
                    e = sys.exc_info()[1]
                    # Some corrupt tar files seem to produce this
                    # (specifically bad symlinks)
                    logger.error('In the tar file %s the member %s is invalid: %s'
                                  % (filename, member.name, e))
                    continue
                if not os.path.exists(os.path.dirname(path)):
                    os.makedirs(os.path.dirname(path))
                destfp = open(path, 'wb')
                try:
                    shutil.copyfileobj(fp, destfp)
                finally:
                    destfp.close()
                fp.close()
                # note: configure ...etc
                os.chmod(path, member.mode)
                # note: the file timestamps should be such that asdl_c.py is not invoked.
                os.utime(path, (member.mtime, member.mtime))
    finally:
        tar.close()

def extract_downloadfile(content_type, download_file, target_dir):
    logger.info("Extracting %s into %s" % (os.path.basename(download_file), target_dir))
    if is_gzip(content_type, download_file):
        untar_file(download_file, target_dir)
        return True
    else:
        logger.error("Cannot determine archive format of %s" % download_file)
        return False

def is_installed(pkg):
    return os.path.isdir(os.path.join(PATH_PYTHONS, pkg.name))

def path_to_fileurl(path):
    path = os.path.normcase(os.path.abspath(path))
    url = urlquote(path)
    url = url.replace(os.path.sep, '/')
    url = url.lstrip('/')
    return 'file:///' + url

def fileurl_to_path(url):
    assert url.startswith('file:'), ('Illegal scheme:%s' % url)
    url = '/' + url[len('file:'):].lstrip('/')
    return urlunquote(url)

def to_str(val):
    if not PY3K:
        # python2
        if isinstance(val, unicode):
            return val.encode("utf-8")
        return val
    if isinstance(val, bytes):
        return val.decode("utf-8")
    return val

def is_str(val):
    if not PY3K:
        # python2
        return isinstance(val, basestring)
    # python3
    return isinstance(val, str)

def is_sequence(val):
    if is_str(val):
        return False
    return (hasattr(val, "__getitem__") or hasattr(val, "__iter__"))


#-----------------------------
# class
#-----------------------------
class ShellCommandException(Exception):
    """General exception during shell command"""

class Subprocess(object):
    def __init__(self, log=None, cwd=None, verbose=False, debug=False):
        self._log = log
        self._cwd = cwd
        self._verbose = verbose
        self._debug = debug

    def chdir(self, cwd):
        self._cwd = cwd

    def shell(self, cmd):
        if self._debug:
            logger.log(cmd)
        if is_sequence(cmd):
            cmd = ''.join(cmd)
        if self._log:
            if self._verbose:
                cmd = "(%s) 2>&1 | tee '%s'" % (cmd, self._log)
            else:
                cmd = "(%s) >> '%s' 2>&1" % (cmd, self._log)
        returncode = subprocess.call(cmd, shell=True, cwd=self._cwd)
        if returncode:
            raise ShellCommandException('%s: failed to `%s`' % (returncode, cmd))

    def call(self, cmd):
        if is_str(cmd):
            cmd = shlex.split(cmd)
        if self._debug:
            logger.log(cmd)

        fp = ((self._log and open(self._log, 'a')) or None)
        p = subprocess.Popen(cmd, stdout=subprocess.PIPE, stderr=subprocess.STDOUT, cwd=self._cwd)
        while p.returncode is None:
            while any(select.select([p.stdout], [], [])):
                line = to_str(p.stdout.readline())
                if not line:
                    break
                if self._verbose:
                    logger.log(line.strip())
                if fp:
                    fp.write(line)
                    fp.flush()
            p.poll()
        if fp:
            fp.close()
        return p.returncode

    def check_call(self, cmd):
        returncode = self.call(cmd)
        if returncode:
            raise ShellCommandException('%s: failed to `%s`' % (returncode, cmd))

class Package(object):
    def __init__(self, version, type):
        type = type.lower()
        if type == 'cpython':
            tag = 'CPython'
        elif type == 'stackless':
            tag = 'Stackless'
        elif type == 'pypy':
            tag = 'PyPy'
        elif type == 'pypy3':
            tag = 'PyPy3'
        elif type == 'jython':
            tag = 'Jython'
        else:
            raise ValueError('invalid type: %s' % type)
        self.type = type
        self.tag = tag
        self.version = version

    @property
    def name(self):
        return '%s-%s' % (self.tag, self.version)

class Link(object):
    def __init__(self, url):
        self._url = url

    @property
    def filename(self):
        url = self._url
        url = url.split('#', 1)[0]
        url = url.split('?', 1)[0]
        url = url.rstrip('/')
        name = posixpath.basename(url)
        assert name, ('URL %r produced no filename' % url)
        return name

    @property
    def base_url(self):
        return posixpath.basename(self._url.split('#', 1)[0].split('?', 1)[0])

class Version(object):
    """version compare
    """
    def __init__(self, v):
        self._version = v
        self._p = self._parse_version(v)

    def __lt__(self, o):
        if is_str(o):
            o = self._parse_version(o)
        return self._p < o

    def __le__(self, o):
        if is_str(o):
            o = self._parse_version(o)
        return self._p <= o

    def __eq__(self, o):
        if is_str(o):
            o = self._parse_version(o)
        return self._p == o

    def __ne__(self, o):
        if is_str(o):
            o = self._parse_version(o)
        return self._p != o

    def __gt__(self, o):
        if is_str(o):
            o = self._parse_version(o)
        return self._p > o

    def __ge__(self, o):
        if is_str(o):
            o = self._parse_version(o)
        return self._p >= o

    def _parse_version(self, s):
        """see pkg_resouce.parse_version
        """
        component_re = re.compile(r'(\d+ | [a-z]+ | \.| -)', re.VERBOSE)
        replace = {'pre':'c', 'preview':'c','-':'final-','rc':'c','dev':'@'}.get

        def _parse_version_parts(s):
            for part in component_re.split(s):
                part = replace(part,part)
                if not part or part=='.':
                    continue
                if part[:1] in '0123456789':
                    yield part.zfill(8)    # pad for numeric comparison
                else:
                    yield '*'+part
            yield '*final'  # ensure that alpha/beta/candidate are before final

        parts = []
        for part in _parse_version_parts(s.lower()):
            if part.startswith('*'):
                if part<'*final':   # remove '-' before a prerelease tag
                    while parts and parts[-1]=='*final-': parts.pop()
                # remove trailing zeros from each series of numeric parts
                while parts and parts[-1]=='00000000':
                    parts.pop()
            parts.append(part)
        return tuple(parts)

    def __repr__(self):
        return self._version
